llvm/include/llvm/Target/GlobalISel/Combine.td

```diff
@@ -365,6 +365,13 @@ def bitreverse_lshr : GICombineRule<
                                                     MRI.getType(${amt}.getReg())}}); }]),
   (apply (G_SHL $d, $val, $amt))>;
 
+def lsb_bits_clearing_by_shifts_to_and: GICombineRule<
+  (defs root:$root, build_fn_matchinfo:$matchinfo),
+  (match (G_LSHR $d, $src, $amt1),
+         (G_SHL $root, $d, $amt2):$root,
```

Comment:
You can put G_CONSTANTs on `amt1` and `amt2`. There are examples in Combine.td.

Reply:
I did but the problem was that there are two case: one where they both use the same G_CONSTANT and the other where they use two different G_CONSTANT, but with the same value. The two different constant case was working but if I added another similar pattern with one and both G_LSHR and G_SHL use that it throwed tablegen errors.

Reply:
```
[7/250] Building AArch64GenPreLegalizeGICombiner.inc...
FAILED: lib/Target/AArch64/AArch64GenPreLegalizeGICombiner.inc /home/david-szabo/work/llvm-community/build/lib/Target/AArch64/AArch64GenPreLegalizeGICombiner.inc 
cd /home/david-szabo/work/llvm-community/build && /home/david-szabo/work/llvm-community/build/NATIVE/bin/llvm-tblgen -gen-global-isel-combiner -combiners="AArch64PreLegalizerCombiner" -I /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AArch64 -I/home/david-szabo/work/llvm-community/build/include -I/home/david-szabo/work/llvm-community/llvm-project/llvm/include -I /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AArch64/AArch64.td --write-if-changed -o lib/Target/AArch64/AArch64GenPreLegalizeGICombiner.inc -d lib/Target/AArch64/AArch64GenPreLegalizeGICombiner.inc.d
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AArch64/AArch64.td:41:
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AArch64/AArch64Combine.td:12:
/home/david-szabo/work/llvm-community/llvm-project/llvm/include/llvm/Target/GlobalISel/Combine.td:368:5: error: Nested instruction '__lsb_bits_clearing_by_shifts_to_and_match_0' cannot be the same as another operand 'amt'
def lsb_bits_clearing_by_shifts_to_and: GICombineRule<
    ^
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AArch64/AArch64.td:41:
/home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AArch64/AArch64Combine.td:68:5: error: Failed to parse one or more rules
def AArch64PreLegalizerCombiner: GICombiner<
    ^
[22/172] Building AMDGPUGenPostLegalizeGICombiner.inc...
FAILED: lib/Target/AMDGPU/AMDGPUGenPostLegalizeGICombiner.inc /home/david-szabo/work/llvm-community/build/lib/Target/AMDGPU/AMDGPUGenPostLegalizeGICombiner.inc 
cd /home/david-szabo/work/llvm-community/build && /home/david-szabo/work/llvm-community/build/NATIVE/bin/llvm-tblgen -gen-global-isel-combiner -combiners="AMDGPUPostLegalizerCombiner" -I /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU -I/home/david-szabo/work/llvm-community/build/include -I/home/david-szabo/work/llvm-community/llvm-project/llvm/include -I /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUGISel.td --write-if-changed -o lib/Target/AMDGPU/AMDGPUGenPostLegalizeGICombiner.inc -d lib/Target/AMDGPU/AMDGPUGenPostLegalizeGICombiner.inc.d
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUGISel.td:14:
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUCombine.td:9:
/home/david-szabo/work/llvm-community/llvm-project/llvm/include/llvm/Target/GlobalISel/Combine.td:368:5: error: Nested instruction '__lsb_bits_clearing_by_shifts_to_and_match_0' cannot be the same as another operand 'amt'
def lsb_bits_clearing_by_shifts_to_and: GICombineRule<
    ^
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUGISel.td:14:
/home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUCombine.td:160:5: error: Failed to parse one or more rules
def AMDGPUPostLegalizerCombiner: GICombiner<
    ^
[23/172] Building AMDGPUGenPreLegalizeGICombiner.inc...
FAILED: lib/Target/AMDGPU/AMDGPUGenPreLegalizeGICombiner.inc /home/david-szabo/work/llvm-community/build/lib/Target/AMDGPU/AMDGPUGenPreLegalizeGICombiner.inc 
cd /home/david-szabo/work/llvm-community/build && /home/david-szabo/work/llvm-community/build/NATIVE/bin/llvm-tblgen -gen-global-isel-combiner -combiners="AMDGPUPreLegalizerCombiner" -I /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU -I/home/david-szabo/work/llvm-community/build/include -I/home/david-szabo/work/llvm-community/llvm-project/llvm/include -I /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUGISel.td --write-if-changed -o lib/Target/AMDGPU/AMDGPUGenPreLegalizeGICombiner.inc -d lib/Target/AMDGPU/AMDGPUGenPreLegalizeGICombiner.inc.d
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUGISel.td:14:
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUCombine.td:9:
/home/david-szabo/work/llvm-community/llvm-project/llvm/include/llvm/Target/GlobalISel/Combine.td:368:5: error: Nested instruction '__lsb_bits_clearing_by_shifts_to_and_match_0' cannot be the same as another operand 'amt'
def lsb_bits_clearing_by_shifts_to_and: GICombineRule<
    ^
Included from /home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUGISel.td:14:
/home/david-szabo/work/llvm-community/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUCombine.td:154:5: error: Failed to parse one or more rules
def AMDGPUPreLegalizerCombiner: GICombiner<
    ^
[24/172] Building AMDGPUGenGlobalISel.inc...
ninja: build stopped: subcommand failed.

```

Reply:
There are 2 constants here, you need to use different names for them

Reply:
```
 (match (G_CONSTANT $amt2, $imm2),
        (G_CONSTANT $amt, $imm),
         (G_LSHR $d, $src, $amt2),
         (G_SHL $root, $d, $amt):$root,
```
May need some tweaking to compile, but gives you an idea.

Reply:
I am a bit suprised this is working for case where there is just a single G_CONSTANT. Now the only problem is that this pattern will not match the vector test case.
```diff
+++ b/llvm/include/llvm/Target/GlobalISel/Combine.td
@@ -367,7 +367,9 @@ def bitreverse_lshr : GICombineRule<
 
 def lsb_bits_clearing_by_shifts_to_and: GICombineRule<
   (defs root:$root, build_fn_matchinfo:$matchinfo),
-  (match (G_LSHR $d, $src, $amt1),
+  (match (G_CONSTANT $amt1, $imm1),
+         (G_CONSTANT $amt2, $imm2),
+         (G_LSHR $d, $src, $amt1),
          (G_SHL $root, $d, $amt2):$root,
          [{ return Helper.matchLsbClearByShifts(*${root}, ${matchinfo}); }]),
   (apply [{ Helper.applyBuildFn(*${root}, ${matchinfo}); }])>;
diff --git a/llvm/test/CodeGen/AArch64/GlobalISel/combine-shifts-to-and.mir b/llvm/test/CodeGen/AArch64/GlobalISel/combine-shifts-to-and.mir
index 71236fb980e3..57c96f9fb9e1 100644
--- a/llvm/test/CodeGen/AArch64/GlobalISel/combine-shifts-to-and.mir
+++ b/llvm/test/CodeGen/AArch64/GlobalISel/combine-shifts-to-and.mir
@@ -124,10 +124,11 @@ body:             |
     ; CHECK-NEXT: [[COPY:%[0-9]+]]:_(p0) = COPY $x0
     ; CHECK-NEXT: [[COPY1:%[0-9]+]]:_(s32) = COPY $w1
     ; CHECK-NEXT: %xvec:_(<4 x s32>) = G_BUILD_VECTOR [[COPY1]](s32), [[COPY1]](s32), [[COPY1]](s32), [[COPY1]](s32)
-    ; CHECK-NEXT: [[C:%[0-9]+]]:_(s32) = G_CONSTANT i32 -4
-    ; CHECK-NEXT: [[BUILD_VECTOR:%[0-9]+]]:_(<4 x s32>) = G_BUILD_VECTOR [[C]](s32), [[C]](s32), [[C]](s32), [[C]](s32)
-    ; CHECK-NEXT: [[AND:%[0-9]+]]:_(<4 x s32>) = G_AND %xvec, [[BUILD_VECTOR]]
-    ; CHECK-NEXT: G_STORE [[AND]](<4 x s32>), [[COPY]](p0) :: (store (<4 x s32>))
+    ; CHECK-NEXT: [[C:%[0-9]+]]:_(s32) = G_CONSTANT i32 2
+    ; CHECK-NEXT: %veccst:_(<4 x s32>) = G_BUILD_VECTOR [[C]](s32), [[C]](s32), [[C]](s32), [[C]](s32)
+    ; CHECK-NEXT: [[LSHR:%[0-9]+]]:_(<4 x s32>) = G_LSHR %xvec, %veccst(<4 x s32>)
+    ; CHECK-NEXT: [[SHL:%[0-9]+]]:_(<4 x s32>) = G_SHL [[LSHR]], %veccst(<4 x s32>)
+    ; CHECK-NEXT: G_STORE [[SHL]](<4 x s32>), [[COPY]](p0) :: (store (<4 x s32>))
     ; CHECK-NEXT: RET_ReallyLR
     %0:_(p0) = COPY $x0
     %1:_(s32) = COPY $w1

```